### ISSUE TITLE
Negative Kmax estimates

### ISSUE THREAD

**Issue by cherfychow** [renatoamorais/rfishprod#3]
Thanks for such a great package, Renato! I have been using the package to calculate fish growth on my data of 579 species in 500 iterations. For 97 length records of 13 species, the modelling's estimated Kmax ≤ 0. For some species, all Kmax estimates are negative, while for others, it's only for certain lengths/SSTs. They're mostly benthic reef-dwellers, but not all benthic reef-dwellers had this issue. I've double checked the traits and input parameters and they all seem realistic and reasonable and well below MaxSize. Should this be possible with the boosted regression trees? Hoping it's user error on my part!

Here is a summary of the records with Kmax ≤ 0. 
```
  Species               Size           MaxSizeTL           Diet      Position 
 Length:97          Min.   :  1.500   Min.   :  9.00   Crus   :54   PelgAs: 0  
 Class :character   1st Qu.:  4.000   1st Qu.:  9.00   Plank  :18   PelgDw: 3  
 Mode  :character   Median :  5.000   Median :  9.00   InvSes : 8   BtPlAs: 0  
                    Mean   :  9.358   Mean   : 33.93   InvMac : 7   BtPlDw: 0  
                    3rd Qu.:  7.000   3rd Qu.: 11.00   Pisc   : 7   BnthAs: 0  
                    Max.   :130.000   Max.   :300.00   InvMic : 2   BnthDw:94  
                                                       (Other): 1              
       a                  b            Genus              Family         
 Min.   :0.000525   Min.   :2.840   Length:97          Length:97         
 1st Qu.:0.014454   1st Qu.:2.990   Class :character   Class :character  
 Median :0.014454   Median :3.070   Mode  :character   Mode  :character  
 Mean   :0.015657   Mean   :3.038                                        
 3rd Qu.:0.015136   3rd Qu.:3.070                                        
 Max.   :0.033113   Max.   :3.270                                        
                                                                         
    sstmean         Method               Kmax             Kmax_lowq        
 Min.   :26.19   Length:97          Min.   :-0.351641   Min.   :-0.351641  
 1st Qu.:26.19   Class :character   1st Qu.:-0.193860   1st Qu.:-0.193860  
 Median :26.68   Mode  :character   Median :-0.016640   Median :-0.016640  
 Mean   :26.89                      Mean   :-0.100372   Mean   :-0.100372  
 3rd Qu.:27.55                      3rd Qu.:-0.016640   3rd Qu.:-0.016640  
 Max.   :28.91                      Max.   :-0.002114   Max.   :-0.002114  
                                                                           
   Kmax_uppq        
 Min.   :-0.351641  
 1st Qu.:-0.193860  
 Median :-0.016640  
 Mean   :-0.100372  
 3rd Qu.:-0.016640  
 Max.   :-0.002114
```

**Comment by renatoamorais** [comment]
Hi Cher,

That should not happen and I never saw it or heard from anybody that got the same. Something seems off with with the prediction because your lower and upper quantiles are exactly the same (unless you specified only 1 iteration).

Can you share data and code to reproduce this?

**Comment by cherfychow** [comment]
Yeah no problem. Thanks for getting back so quickly. Here's the code and data: 
[negKmax_forRenato.zip](https://github.com/user-attachments/files/16708782/negKmax_forRenato.zip)

**Comment by helenfyan** [comment]
@renatoamorais I think you need to set objective = 'reg:gamma' in your xgboost() function to limit Kmax to positive-only values

**Comment by renatoamorais** [comment]
Geeeez, sorry, completely blanked out and forgot about this! It comes to my personal email which I don't check as often as I (presumably) should, so reminders went through unnoticed...

@cherfychow, I cannot reproduce your error because your input data didn't come with the file you sent. The only data I managed to find there ('vbgf_traits_estimates_forRenato.csv') does not have any negative values. 

@helenfyan all xgboost params are loaded automatically, including objective = 'reg:gamma', so that should in principle not be the problem. @cherfychow can you please try manually specifying the parameters as per the code below if you still have this problem?

```
mod_params <- rfishprod:::xgboostparams
preds <- predKmax(traits, db, fmod, params = mod_params)

```

**Comment by cherfychow** [comment]
No worries! Thank you for helping me out with this. It took me a while to be able to try this out again on my full data with 1000 iterations. I tried your suggestion and made sure gamma was specified and still ended up with a small minority of negative predictions. I've attached the pred table of just the negative records if you have any ideas where things might've gone wrong (the input trait table is just the first 14 columns). I'm going to attempt optimising the parameters with xgb.train in the meantime to see if that helps.

[negK.csv](https://github.com/user-attachments/files/19051294/negK.csv)

**Comment by cherfychow** [comment]
Sorry! took me a while to do a full rerun with all my data. I don't know what exactly it was, but I think I was having a mix of needing to really specify the arguments to R and running on a high powered local machine that did the trick. Thanks for all the help!

**Comment by cherfychow** [comment]
An update: after some trialling, I think the negative Kmax predicted values and zero confidence interval happens when the test data is too large for xgboost to converge (or the xgboost equivalent to that?). At the very upper limit, R will give some error about a vector limit, but prediction failure happens way before that. Neither R or the packages throw warning or error messages that hint at something going wrong so I thought you'd like to know.

**Comment by cherfychow** [comment]
Sorry please disregard the last update. I dug a little deeper under the hood into the `predKmax` function. My R session did crash if the test data was too large. But ultimately, your initial suspicion about the parameters not being fed in was partially true. I found out that if I modified the training data, the function passes an empty list instead of the default parameters or parameters defined in arguments. The runs without parameters defined was the main culprit in negative values and zero confidence intervals. It was mostly my error, but maybe a warning is needed in documentation.
```
if(identical (dataset, db) & is.null (params)) {
	params <- xgboostparams
} else {
	params <- list ()
	warning('Consider optimising xgboost parameters with xgb.train')
}
```

**Comment by renatoamorais** [comment]
This issue has now been officially solved. It involved the argument objective not being passed correctly from predKmax to xgboost because of an update that changed the way in which arguments are specified. All should be fine now.
